drivers/ipm/ipm_adsp.c

```diff
@@ -0,0 +1,212 @@
+
+/* FIXME define in DTS */
+#define DT_IPM_ADSP_0_BASE_ADDRESS	0x10
+#define DT_IPM_ADSP_0_NAME		"IPM_0"
+#define DT_IPM_ADSP_0_IRQ		SOC_AGGREGATE_IRQ(7, 6)
+#define DT_IPM_ADSP_0_IRQ_PRI		3
+
+/* FIXME: set to 64 for now */
+#define IPM_ADSP_MAX_DATA_SIZE		256
+#define IPM_ADSP_MAX_ID_VAL		IPC_DIPCI_MSG_MASK
+
+/* Mailbox ADSP -> Host */
+#define IPM_ADSP_MAILBOX_OUT		MAILBOX_DSPBOX_BASE
+#define IPM_ADSP_MAILBOX_OUT_SIZE	MAILBOX_DSPBOX_SIZE
+
+/* Mailbox Host -> ADSP */
+#define IPM_ADSP_MAILBOX_IN		MAILBOX_HOSTBOX_BASE
+#define IPM_ADSP_MAILBOX_IN_SIZE	MAILBOX_HOSTBOX_SIZE
+
+struct ipm_adsp_config {
+	u8_t *base;
+	void (*irq_config_func)(struct device *dev);
+};
+
+struct ipm_adsp_data {
+	ipm_callback_t callback;
+	void *callback_ctx;
+};
+
+static void ipm_adsp_isr(void *arg)
+{
+	struct device *dev = (struct device *)arg;
+	const struct ipm_adsp_data *data = dev->driver_data;
+	uint32_t dipcctl, dipcie, dipct;
+
+	dipct = ipc_read(IPC_DIPCT);
+	dipcie = ipc_read(IPC_DIPCIE);
+	dipcctl = ipc_read(IPC_DIPCCTL);
+
+	LOG_DBG("dipct 0x%x dipcie 0x%x dipcctl 0x%x", dipct, dipcie, dipcctl);
+
+	/* IPC Target Busy Interrupt Enable (IPCTBIE) */
+	if (dipct & IPC_DIPCT_BUSY && dipcctl & IPC_DIPCCTL_IPCTBIE) {
+		/* Mask Busy interrupt */
+		ipc_write(IPC_DIPCCTL, dipcctl & ~IPC_DIPCCTL_IPCTBIE);
+
+		if (data->callback) {
+			SOC_DCACHE_INVALIDATE((void *)IPM_ADSP_MAILBOX_IN,
+					      IPM_ADSP_MAILBOX_IN_SIZE);
+			/* Use zero copy */
+			data->callback(data->callback_ctx,
+				       dipct & IPC_DIPCI_MSG_MASK,
+				       (void *)IPM_ADSP_MAILBOX_IN);
+		}
+
+		/* Clear BUSY enabling interrupts */
+		ipc_write(IPC_DIPCT, ipc_read(IPC_DIPCT) | IPC_DIPCT_BUSY);
+
+		/* Unmask interrupts */
+		ipc_write(IPC_DIPCCTL,
+			  ipc_read(IPC_DIPCCTL) | IPC_DIPCCTL_IPCTBIE);
+	}
+
+	if (dipcie & IPC_DIPCIE_DONE && dipcctl & IPC_DIPCCTL_IPCIDIE) {
+		/* Mask Done interrupt */
+		ipc_write(IPC_DIPCCTL,
+			  ipc_read(IPC_DIPCCTL) & ~IPC_DIPCCTL_IPCIDIE);
+
+		/* Clear DONE bit - tell host we have completed the operation */
+		ipc_write(IPC_DIPCIE,
+			  ipc_read(IPC_DIPCIE) | IPC_DIPCIE_DONE);
+
+		/* Unmask Done interrupt */
+		ipc_write(IPC_DIPCCTL,
+			  ipc_read(IPC_DIPCCTL) | IPC_DIPCCTL_IPCIDIE);
+
+		LOG_DBG("Not handled: IPC_DIPCCTL_IPCIDIE");
+
+		/* TODO: implement queued message sending if needed */
+	}
+}
+
+static int ipm_adsp_send(struct device *dev, int wait, u32_t id,
+			 const void *data, int size)
+{
+	LOG_DBG("Send: id %d data %p size %d", id, data, size);
+	LOG_HEXDUMP_DBG(data, size, "send");
+
+	if (id > IPM_ADSP_MAX_ID_VAL) {
+		return -EINVAL;
+	}
+
+	if (size > IPM_ADSP_MAX_DATA_SIZE) {
+		return -EMSGSIZE;
+	}
+
+	if (wait) {
+		while(ipc_read(IPC_DIPCI) & IPC_DIPCI_BUSY) {
+		}
+	} else {
+		if (ipc_read(IPC_DIPCI) & IPC_DIPCI_BUSY) {
+			LOG_DBG("Busy: previous message is not handled");
+			return -EBUSY;
+		}
+	}
+
+	memcpy((void *)IPM_ADSP_MAILBOX_OUT, data, size);
+	SOC_DCACHE_FLUSH((void *)IPM_ADSP_MAILBOX_OUT, size);
```

Comment:
Given that this is an app-visible API, you almost certainly want to check size vs. MAILBOX_OUT_SIZE to avoid scribbling over shared areas and potentially confusing/crashing the host kernel.  Or at the very least check it in an __ASSERT().

So this supports only one message at a time?  A quick grep says that the outbox region is 4k, I think, so it seems like some queuing ability would really improve things.

Reply:
About queuing. API mentions:
https://github.com/zephyrproject-rtos/zephyr/blob/bd5048c251bd578776d063ac45a94f8b482c5ed4/include/drivers/ipm.h#L118

Reply:
> Or at the very least check it in an __ASSERT().

Added `BUILD_ASSERT`.